v2/tools/generator/internal/codegen/pipeline/add_cross_resource_references.go

```diff
@@ -243,6 +106,92 @@ func makeLegacyReferencePropertyName(existing *astmodel.PropertyDefinition, isSl
 	return referencePropertyName
 }
 
+type ARMIDToGenruntimeReferenceTypeVisitor struct {
+	astmodel.TypeVisitor
+	idFactory astmodel.IdentifierFactory
+}
+
+func (v *ARMIDToGenruntimeReferenceTypeVisitor) transformARMIDToGenruntimeResourceReference(_ *astmodel.TypeVisitor, it *astmodel.PrimitiveType, _ interface{}) (astmodel.Type, error) {
+	if it == astmodel.ARMIDType {
+		return astmodel.ResourceReferenceType, nil
+	}
+
+	return it, nil
+}
+
+func (v *ARMIDToGenruntimeReferenceTypeVisitor) renamePropertiesWithARMIDReferences(this *astmodel.TypeVisitor, it *astmodel.ObjectType, ctx interface{}) (astmodel.Type, error) {
+	// First, we visit this type like normal
+	result, err := astmodel.IdentityVisitOfObjectType(this, it, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	typeName := ctx.(astmodel.TypeName)
+	ot, ok := result.(*astmodel.ObjectType)
+	if !ok {
+		return nil, errors.Errorf("somehow result for visitObjectType of %s was not objectType, instead %T", typeName, result)
```

Comment:
Suggest this should be more positive (negative?) about the error.

```suggestion
		return nil, errors.Errorf("result for visitObjectType of %s was not expected ObjectType, instead %T", typeName, result)
```

Reply:
Agreed. Fixed